include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce_then_scan.h

```diff
@@ -0,0 +1,835 @@
+    __sub_group_size, __max_inputs_per_item, __is_inclusive, _GenReduceInput, _ReduceOp, _GenScanInput,
+    _ScanInputTransform, _WriteOp, _InitType, __internal::__optional_kernel_name<_KernelName...>>
+{
+
+    template <typename _TmpPtr>
+    auto
+    __get_block_carry_in(const std::size_t __block_num, _TmpPtr __tmp_ptr) const
+    {
+        return __tmp_ptr[__num_sub_groups_global + (__block_num % 2)];
+    }
+
+    template <typename _TmpPtr, typename _ValueType>
+    void
+    __set_block_carry_out(const std::size_t __block_num, _TmpPtr __tmp_ptr, const _ValueType __block_carry_out) const
+    {
+        __tmp_ptr[__num_sub_groups_global + 1 - (__block_num % 2)] = __block_carry_out;
+    }
+
+    template <typename _ExecutionPolicy, typename _InRng, typename _OutRng, typename _TmpStorageAcc>
+    auto
+    operator()(_ExecutionPolicy&& __exec, const sycl::nd_range<1> __nd_range, _InRng&& __in_rng, _OutRng&& __out_rng,
+               _TmpStorageAcc __scratch_container, const sycl::event& __prior_event,
+               const std::size_t __inputs_per_sub_group, const std::size_t __inputs_per_item,
+               const std::size_t __block_num) const
+    {
+        std::size_t __elements_in_block = std::min(__n - __block_num * __max_block_size, std::size_t(__max_block_size));
+        std::size_t __active_groups = oneapi::dpl::__internal::__dpl_ceiling_div(
+            __elements_in_block, __inputs_per_sub_group * __num_sub_groups_local);
+        using _InitValueType = typename _InitType::__value_type;
+        return __exec.queue().submit([&, this](sycl::handler& __cgh) {
+            __dpl_sycl::__local_accessor<_InitValueType> __sub_group_partials(__num_sub_groups_local + 1, __cgh);
+            __cgh.depends_on(__prior_event);
+            oneapi::dpl::__ranges::__require_access(__cgh, __in_rng, __out_rng);
+            auto __temp_acc = __scratch_container.__get_scratch_acc(__cgh);
+            auto __res_acc = __scratch_container.__get_result_acc(__cgh);
+
+            __cgh.parallel_for<_KernelName...>(__nd_range, [=,
+                                                            *this](sycl::nd_item<1> __ndi) [[sycl::reqd_sub_group_size(
+                                                               __sub_group_size)]] {
+                auto __tmp_ptr = _TmpStorageAcc::__get_usm_or_buffer_accessor_ptr(__temp_acc);
+                auto __res_ptr =
+                    _TmpStorageAcc::__get_usm_or_buffer_accessor_ptr(__res_acc, __num_sub_groups_global + 2);
+                auto __lid = __ndi.get_local_id(0);
+                auto __g = __ndi.get_group(0);
+                auto __sub_group = __ndi.get_sub_group();
+                auto __sub_group_id = __sub_group.get_group_linear_id();
+                auto __sub_group_local_id = __sub_group.get_local_linear_id();
+
+                auto __group_start_idx =
+                    (__block_num * __max_block_size) + (__g * __inputs_per_sub_group * __num_sub_groups_local);
```

Comment:
Similar comment about adding a helper for `__inputs_per_sub_group * __num_sub_groups_local` and reusing it on the next line.